File: day_19/p2.py
def parse(fname: str):
    lines = []
    with open(fname, 'r') as f:
        lines = [line for line in f.read().splitlines() if line]
    
    towels = lines[0].split(', ')
    designs = lines[1:]
    
    return towels, designs

def dp(target, towels, memo):
    if target in memo.keys():
        return memo[target]

    if len(target) == 0:
        return 1
    
    num = 0
    for t in towels:
        if target.startswith(t):
            num += dp(target[len(t):], towels, memo)

    memo[target] = num
    return num

def compute(towels, designs):
    count = 0
    for d in designs:
        memo = {}
        count += dp(d, towels, memo)
    
    return count

def main():
    towels, designs = parse("input.txt")
    print(compute(towels, designs))


if __name__ == "__main__":
    main()

File: day_19/p2.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>

std::pair<strings, strings> Parse(const std::string& fname)
{
    strings towels {};
    strings designs {};
    auto lines = pypp::splitFileLines(fname);

    for (int i { 0 }; i < lines.size(); ++i) {
        if (i == 0)
            towels = pypp::split(lines[i], ',');
        else
            designs.push_back(lines[i]);
    }

    for (auto& t : towels)
        t = pypp::strip(t, [](unsigned char ch) { return std::isspace(ch); });

    return { towels, designs };
}

long dp(const std::string& target, const strings& towels, std::unordered_map<std::string, long>& memo)
{
    if (memo.contains(target))
        return memo[target];

    if (target.length() == 0)
        return 1;

    long num { 0L };
    for (auto t : towels)
        if (target.find(t) == 0)
            num += dp(target.substr(t.length()), towels, memo);

    memo[target] = num;
    return num;
}

long Compute(const strings& towels, const strings& designs)
{
    long count { 0 };

    for (auto d : designs) {
        std::unordered_map<std::string, long> memo;
        count += dp(d, towels, memo);
    }

    return count;
}

int main()
{
    const auto [towels, designs] = Parse("input.txt");
    std::cout << Compute(towels, designs) << std::endl;

    return 0;
}
